### PR TITLE
Lot's of small improvements for ConstraintSolver

### PR DESCRIPTION
**Pull request by forki**

### CODE REVIEW

File: src/fsharp/ConstraintSolver.fsi [hunk 1]
```diff
@@ -94,7 +94,7 @@ type ConstraintSolverState =
 
 type ConstraintSolverEnv 
 
-val BakedInTraitConstraintNames : string list
+val BakedInTraitConstraintNames : Set<string>
```
Comment (smoothdeveloper) [inline src/fsharp/ConstraintSolver.fsi]: Nice but how does it affect performance? I assume insertion might cost more in Set.

File: src/fsharp/ConstraintSolver.fs [hunk 2]
```diff
@@ -1417,10 +1420,10 @@ and SolveRelevantMemberConstraints (csenv:ConstraintSolverEnv) ndeep permitWeakR
             tps |> AtLeastOneD (fun tp -> 
                 /// Normalize the typar 
                 let ty = mkTyparTy tp
-                if isAnyParTy csenv.g ty then 
-                    let tp = destAnyParTy csenv.g ty
+                match tryAnyParTy csenv.g ty with
```
Comment (smoothdeveloper) [inline src/fsharp/ConstraintSolver.fs]: you have done similar changes in separate PR, is it a pattern that keeps happening in code base: one function call for branch, followed by another call, always the two same functions?

**Comment by dsyme** [comment]
LGTM.  Could probably use a HashSet rather than a (tree) Set, the HashSet is faster

**Comment by forki** [comment]
rebased and fixed

File: src/fsharp/ConstraintSolver.fs [hunk 3]
```diff
@@ -944,14 +943,13 @@ and SolveMemberConstraint (csenv:ConstraintSolverEnv) ignoreUnresolvedOverload p
...
     
-    // Assert the object type if the constraint is for an instance member
-    begin
-        if memFlags.IsInstance then 
-            match tys, argtys with 
-            | [ty], (h :: _) -> SolveTypEqualsTypKeepAbbrevs csenv ndeep m2 trace h ty
-            | _ -> ErrorD (ConstraintSolverError(FSComp.SR.csExpectedArguments(), m,m2))
-        else CompleteD
-    end ++ (fun () -> 
+    // Assert the object type if the constraint is for an instance member
```
Comment (dsyme) [inline src/fsharp/ConstraintSolver.fs]: I prefer the original begin/end please as then there is no doubt that it's ``(if A then B else C) ++ (...)`` rather than ``if A then B else (C + (...))``

File: src/fsharp/ConstraintSolver.fs [hunk 4]
```diff
@@ -1552,43 +1542,42 @@ and AddConstraint (csenv:ConstraintSolverEnv) ndeep m2 trace tp newConstraint  =
...
-
-        enforceMutualConsistency 0 allCxs 
-    end ++ (fun ()  ->
+     
+    let rec enforceMutualConsistency i cxs = 
+        match cxs with 
+        | [] ->  CompleteD
+        | cx :: rest -> IterateIdxD (fun j cx2 -> if i = j then CompleteD else consistent cx cx2) allCxs ++ (fun () -> enforceMutualConsistency (i+1) rest)
+
+    enforceMutualConsistency 0 allCxs
```
Comment (dsyme) [inline src/fsharp/ConstraintSolver.fs]: Again explicit parentheses to clarify ``(a @ b) ++ c`` rather than ``a @ b ++ c`` independent of whitespace rules please

File: src/fsharp/ConstraintSolver.fs (hunk 3 above)
Comment (forki) [inline src/fsharp/ConstraintSolver.fs]: done

File: src/fsharp/ConstraintSolver.fs (hunk 4 above)
Comment (forki) [inline src/fsharp/ConstraintSolver.fs]: done

File: src/fsharp/TypeChecker.fs [hunk 5]
```diff
@@ -5977,7 +5979,7 @@ and TcExprUndelayed cenv overallTy env tpenv (expr: SynExpr) =
     | SynExpr.TraitCall(tps,memSpfn,arg,m) ->
         let synTypes =  tps |> List.map (fun tp -> SynType.Var(tp,m))
         let (TTrait(_,logicalCompiledName,_,argtys,returnTy,_) as traitInfo),tpenv = TcPseudoMemberSpec cenv NewTyparsOK env synTypes tpenv memSpfn m
-        if List.contains logicalCompiledName BakedInTraitConstraintNames then 
+        if BakedInTraitConstraintNames.Contains logicalCompiledName then
```
Comment (KevinRansom) [inline src/fsharp/TypeChecker.fs]: Do we know that this is only called single threaded?  perhaps ConcurrentBag or ConcurrentDictionary?

File: src/fsharp/ConstraintSolver.fsi [hunk 6]
```diff
@@ -94,7 +94,7 @@ type ConstraintSolverState =
 
 type ConstraintSolverEnv 
 
-val BakedInTraitConstraintNames : string list
+val BakedInTraitConstraintNames : System.Collections.Generic.HashSet<string>
```
Comment (KevinRansom) [inline src/fsharp/ConstraintSolver.fsi]: Do we know that this is only used single threaded?  How do we guarantee that moving forward?

File: src/fsharp/TypeChecker.fs (hunk 5 above)
Comment (dsyme) [inline src/fsharp/TypeChecker.fs]: Practically speaking I'm pretty certain HashSet is safe when used as an immutable collection (concurrent readers with no further writers).  I also suspect we rely on that kind of property elsewhere (i.e. "mutable collection becomes threadsafe after last write") 

But it's true that the documentation doesn't say that specifically...  Maybe use an System.Collections.Immutable collection explicitly.

Comment (forki) [inline src/fsharp/TypeChecker.fs]: Ok went back to set. I think this is still improvement over plain list.

File: src/fsharp/ConstraintSolver.fs [hunk 7]
```diff
@@ -1933,8 +1933,8 @@ and CanMemberSigsMatchUpToCheck
         // - Never take into account return type information for constructors 
         match reqdRetTyOpt with 
         | None -> CompleteD 
-        | Some _  when minfo.IsConstructor -> CompleteD 
-        | Some _  when not alwaysCheckReturn && isNil unnamedCalledOutArgs -> CompleteD 
+        | _ when minfo.IsConstructor -> CompleteD 
+        | _ when not alwaysCheckReturn && isNil unnamedCalledOutArgs -> CompleteD
```
Comment (dsyme) [inline src/fsharp/ConstraintSolver.fs]: I actually prefer leaving the ``Some _`` in in this case, I don't have to compute it for myself and it makes the clauses reorderable

Comment (forki) [inline src/fsharp/ConstraintSolver.fs]: done

**Comment by smoothdeveloper** [comment]
@forki when you have settled on your changes, do you mind updating the PR description about intended changes beyond the stylistic ones?

AFAIU, only the change to Set instead of List which was used for lookup is the only change, is there other things? Is it expected to save some compile time or more like a "don't do stupid things in compiler"?

Just wondering and I think the stylistic cleanup is a good effort along the way.

**Comment by forki** [comment]
ok I'm done for now

**Comment by KevinRansom** [comment]
Thanks for this

Kevin
